README.md
# ember-aupac-typeahead

[![NPM package](https://img.shields.io/npm/v/ember-aupac-typeahead.svg)](https://www.npmjs.com/package/ember-aupac-typeahead) [![Build Status](https://img.shields.io/travis/aupac/ember-aupac-typeahead.svg)](https://travis-ci.org/aupac/ember-aupac-typeahead) [![Ember Observer Score](http://emberobserver.com/badges/ember-aupac-typeahead.svg)](http://emberobserver.com/addons/ember-aupac-typeahead)

![alt tag](https://github.com/aupac/ember-aupac-typeahead/blob/master/example.jpg)

A [typeahead.js](https://twitter.github.io/typeahead.js/) autocomplete for ember.
Updated to work with maintained fork [corejs-typeahead](https://github.com/corejavascript/typeahead.js))
 - supports any datasource let it be local or remote.
 - explicit support for ember-data via the `aupac-ember-data-typeahead` component.
 - very customisable to meet your needs.
 
Demo [HERE](http://aupac.github.io/ember-aupac-typeahead/)

## Installation

`ember install ember-aupac-typeahead`

## Prerequisites

 * `ember-data` > 1.13.x if using the `aupac-ember-data-typeahead` component
 * `ember.js` > 1.13.x

## `aupac-ember-data-typeahead` component
The `aupac-ember-data-typeahead` component is an extension of the more generic `aupac-typeahead` and assumes you're using ember-data to retrieve your data remotely. This allows ember-data users to streamline the use of this component into a single line of code in their template.

#### Component Attributes
By default, each ember-data model supplied in `modelClass` is required to have a `displayName` (computed property or attribute) that will return a string representing the name to display in the suggestion template.  If this is not possible you can override the `suggestionTemplate` and supply something else (see below).

In addition to all the features supported by `aupac-typeahead` (see below), `aupac-ember-data-typeahead` supports the following:

-  `modelClass` : (*required) the dasherized form of the ember-data model you're searching for. ie 'customer-address'
-  `displayKey` : (default: 'displayName') the attribute to display to the user when an item is selected,
-  `params` : (default: {}) an object containing various query string parameters to send along with the remote request,
-  `queryKey`: (default: 'q') the query parameter sent to the server containing the search text.
-  `selection` : (default: null) initial selection - can be an `ember-data` model (in which case the `displayKey` is used as the initial value) or a `string` which will display as is. Wrap selection in `(readonly x)` helper to avoid two-way binding.

This component has already implemented the relevant functions to make them compatible with ember-data.  You do not need to do so yourself.

#### Usage example
```html
<!--In this case the ember-data model "task" needs a displayName attribute-->
{{aupac-ember-data-typeahead modelClass='task' action=(action (mut selection))}}
```

The above is all you need to have a fully functional autocomplete search in your page.  It would create an input that allows you to search for tasks and when selected would update the `selection` property on your controller.

## `aupac-typeahead` component
The `aupac-typeahead` component contains no assumptions about how you're retrieving your data.  Both local and remote suggestions are supported.

#### Component Attributes
-  `disabled` : (deafult: false) true if the control should be disabled.
-  `placeholder` : (default: 'Search') the placeholder text to display in the input.
-  `name` : (default: '') the name of the typeahead input.
-  `action`: (*required) the selected item will be provided as the first argument.
-  `selection` : (default: null) will be set as the initial selection in the component. Wrap selection with helper `(readonly x)` to avoid two-way binding. 
-  `autoFocus`: (default: false) focus the control on render.
-  `transformSelection`: (default: no transform) allows you to transform the selected value before it is set on the typeahead by returning the transformed value, signature `function(selection)`
-  `allowFreeInput`: (default: false) allows the user to input their own values that are not part of the option list.  Only useful if the item being selected is a String.
-  `tabindex`: allows you to define a numeric tab index for the input

See the [typeahead docs](https://github.com/twitter/typeahead.js/blob/master/doc/jquery_typeahead.md#datasets) for a more complete description of the items below.
-  `source` : (*required) a function to return an array of items to display to the user with the signature `function(query, syncResults, asyncResults)`. The callback functions `syncResults or asyncResults` should be called with an array of results as a parameter.
-  `async` : (default: false) true if the returned data is asynchronous.
-  `datasetName` : (default: 'default') the name of the dataset.
-  `limit` : (default: 15) the maximum number of results to display to the user.
-  `display` : (default: will display the returned item as is) function that displays the selected item to the user, signature `function(model)`.
-  `suggestionTemplate` : a precompiled HTMLBars template used for suggestions, attribute bindings should be specified under the model object. ie `{{model.firstName}}`. If the returned value is **not** an object, it will be bound under {{model.displayName}}.
-  `notFoundTemplate` : a precompiled HTMLBars template that is rendered when no results are found.
-  `pendingTemplate` :  a precompiled HTMLBars template that is rendered when loading the result set but not yet resolved.
-  `headerTemplate` : a precompiled HTMLBars template displayed at the top of the search results.
-  `footerTemplate` : a precompiled HTMLBars template displayed at the bottom of the search results.

See the [typeahead docs](https://github.com/twitter/typeahead.js/blob/master/doc/jquery_typeahead.md#options) for a more complete description of the items below.
-  `highlight`: (default: true) true if matching text be highlighted in the search results.
-  `hint`: (default: true) true if hints be displayed in the input.
-  `minLength`: (default: 2) the minumum number of characters before a search in performed.
-  `typeaheadClassNames`: (default: {}) allows you to customise the class names used in typeahead.

##### Example
In your template
```javascript
{{aupac-typeahead action=(action (mut country))
  class='form-control'
  source=countrySource
  placeholder='Search for a country'}}
```

In your controller
```javascript
const countries = Ember.A(["Afghanistan","Albania","Algeria","Andorra","Angola","Anguilla","Antigua &amp; Barbuda","Argentina","Armenia","Aruba","Australia","Austria","Azerbaijan","Bahamas"
  ,"Bahrain","Bangladesh","Barbados","Belarus","Belgium","Belize","Benin","Bermuda","Bhutan","Bolivia","Bosnia &amp; Herzegovina","Botswana","Brazil","British Virgin Islands"
  ,"Brunei","Bulgaria","Burkina Faso","Burundi","Cambodia","Cameroon","Cape Verde","Cayman Islands","Chad","Chile","China","Colombia","Congo","Cook Islands","Costa Rica"
  ,"Cote D Ivoire","Croatia","Cruise Ship","Cuba","Cyprus","Czech Republic","Denmark","Djibouti","Dominica","Dominican Republic","Ecuador","Egypt","El Salvador","Equatorial Guinea"
  ,"Estonia","Ethiopia","Falkland Islands","Faroe Islands","Fiji","Finland","France","French Polynesia","French West Indies","Gabon","Gambia","Georgia","Germany","Ghana"
  ,"Gibraltar","Greece","Greenland","Grenada","Guam","Guatemala","Guernsey","Guinea","Guinea Bissau","Guyana","Haiti","Honduras","Hong Kong","Hungary","Iceland","India"
  ,"Indonesia","Iran","Iraq","Ireland","Isle of Man","Israel","Italy","Jamaica","Japan","Jersey","Jordan","Kazakhstan","Kenya","Kuwait","Kyrgyz Republic","Laos","Latvia"
  ,"Lebanon","Lesotho","Liberia","Libya","Liechtenstein","Lithuania","Luxembourg","Macau","Macedonia","Madagascar","Malawi","Malaysia","Maldives","Mali","Malta","Mauritania"
  ,"Mauritius","Mexico","Moldova","Monaco","Mongolia","Montenegro","Montserrat","Morocco","Mozambique","Namibia","Nepal","Netherlands","Netherlands Antilles","New Caledonia"
  ,"New Zealand","Nicaragua","Niger","Nigeria","Norway","Oman","Pakistan","Palestine","Panama","Papua New Guinea","Paraguay","Peru","Philippines","Poland","Portugal"
  ,"Puerto Rico","Qatar","Reunion","Romania","Russia","Rwanda","Saint Pierre &amp; Miquelon","Samoa","San Marino","Satellite","Saudi Arabia","Senegal","Serbia","Seychelles"
  ,"Sierra Leone","Singapore","Slovakia","Slovenia","South Africa","South Korea","Spain","Sri Lanka","St Kitts &amp; Nevis","St Lucia","St Vincent","St. Lucia","Sudan"
  ,"Suriname","Swaziland","Sweden","Switzerland","Syria","Taiwan","Tajikistan","Tanzania","Thailand","Timor L'Este","Togo","Tonga","Trinidad &amp; Tobago","Tunisia"
  ,"Turkey","Turkmenistan","Turks &amp; Caicos","Uganda","Ukraine","United Arab Emirates","United Kingdom","Uruguay","Uzbekistan","Venezuela","Vietnam","Virgin Islands (US)"
  ,"Yemen","Zambia","Zimbabwe"]);

export default Ember.Controller.extend({

  country : null,
  countrySource : function(query, syncResults, asyncResults) {
    const regex = new RegExp(`.*${query}.*`, 'i');
    const results = countries.filter((item, index, enumerable) => {
      return regex.test(item);
    })
    syncResults(results);
  }

});
```

## Using your own custom template
You can override the `suggestionTemplate`, `notFoundTemplate`, `pendingTemplate`, `headerTemplate` or `footerTemplate` used by importing a `*.hbs` file and assigning to the appropriate property.

For example
```handlebars
{{!-- app/templates/country-templates/suggestion.hbs --}}
<div class='typeahead-suggestion'><img src="http://www.gravatar.com/avatar/0cf15665a9146ba852bf042b0652780a?s=200" style="width: 10%; height: 10%">{{model.displayName}}</div>
```

Then in your controller
```javascript
import customSuggestionTemplate from '../templates/country-templates/suggestion';

export default Ember.Controller.extend({

  customSuggestionTemplate: customSuggestionTemplate

})
```

And assign it to your template
```handlebars
{{aupac-typeahead action=(action (mut country))
... bind the custom suggestion template to the component
suggestionTemplate=customSuggestionTemplate
}}
````

## Using your own version of [typeahead.js](https://twitter.github.io/typeahead.js/)
You can disable the importing of typeahead.js by adding the following to your `/config/environment.js`

```javascript
'ember-aupac-typeahead' : {
  includeTypeahead: false
}
```

The current compatible typeahead.js version is *v0.11.1*

## CSS Styling
By default, Bootstrap 3 compatible css styles are included with the addon, you can disable this by adding:

```javascript
'ember-aupac-typeahead' : {
  includeCss: false
}
```

See the [typeahead.js docs](https://github.com/twitter/typeahead.js/blob/master/doc/jquery_typeahead.md#class-names) for applying your own custom styling.

## Testing

#### [ember-cli-page-object](https://github.com/san650/ember-cli-page-object) is supported

`test/pages/aupac-typeahead.js`

```javascript
export function typeahead(selector, options) {
  return {
    search : function(search) {
      $(selector).val(search).trigger('input');
    },
    suggestions : collection({
      scope: '', //Reset to global scope
      itemScope: '.tt-suggestion',
      item: {
        select: clickable()
      }
    })
  };
}
```

TODO - show example


## Running

* `ember server`
* Visit your app at http://localhost:4200.

## Running Tests

* `ember test`
* `ember test --server`

## Building

* `ember build`

For more information on using ember-cli, visit [http://www.ember-cli.com/](http://www.ember-cli.com/).


CHANGELOG.md
# Master

# 3.1.0
* (enhancement) Fixes behavior in Ember 3.1

# 3.0.0
* (enhancement) Updated to support ember@2.17.0
* (breaking) Now using npm instead of bower for corejs-typeahead

# 2.3.3
* (bug) #63 Action not called when value cleared out

# 2.3.2
* (enhancement) - send event name to action

# 2.3.1

* (bug) #59 Fix called method

# 2.3.0

* #58 Fix assetPath in treeForVendor
* #57 Remove unnecessary calls to action when focusing out
* #54 Fix minor typo and remove extra whitespace in README
* #53 (breaking) Prepare for ember-cli-fastboot 1.0 build format

  Please update to ember-fastboot > 1.0

* #49 Add syntax highlighting to README
* #48 Fix binding of includeCss and includeTypeahead, add Fastboot support

# 2.2.0

* #46 updates unmaintained typeahead.js dependency to use maintained fork core-typeahead
* #44 upgrade dependencies

# 2.1.2

* #37 (improvement) - Remove deprecated Ember.K

# 2.1.1
* #35 (breaking) - add support for ember@2.10.0 when model is a string

All templates `footerTemplate`, `headerTemplate`, `notFoundTemplate`, `pendingTemplate` and `suggestionTemplate` are now supplied an object.

`suggestionTemplate`: is now always using displayName event when the returned result is a String.
{{model.diplayName}} whereas before it would have just been {{model}}

All other templates were previously passed a `query` or `suggestions` or both directly.
Now these are all prefixed with "model".

So `{{model.query}}` instead of `{{query}}` directly.

* If you are not using custom templates you will not be affected.
* If you were using aupac-ember-data-typeahead and customised the suggestion template you will not be affected.

# 2.1.0
* #31 (enhancement) - Make compatible with ember@2.10.0, fixes submit issue
* #28 (bug) - Fix test helper

# 2.0.8
* #25 (enhancement) - Add tabIndex property
* #23 (enhancement) - Update for compatibility with ember 2.8.0
* #22 (enhancement) - Fixed test helper and added tests.
* #17 (enhancement) - transformSelection is now a computed property.

# 2.0.7
* #17 (bug) properly cleanup event handlers

# 2.0.6
* #15 (enhancement) - Use Ember.$ instead of $

# 2.0.5
* #13 (enhancement) pressing the enter key now behaves the same and focusout.
                    overriding the setValue(selection) function is now deprecated in favour of returning a value from transformSelection(selection)

# 2.0.4
* #11 (enhancement) new allowFreeInput option that will not clear the text on focus out.

# 2.0.3
* #6 (enhancement) added CHANGELOG.md
* #9 **BREAKING** "Maximum call stack size exceeded" due to computed.reads('selection') in ember 2.2.  This addon now requires the `readonly` helper on the selection `selection=(readonly model)` which is a simple mechanical refactor.

# 2.0.2
* #3 (regression) setting selection to null not working

# 2.0.1
* #5 (bug) toggling disabled property breaks css

# 2.0.0
* **BREAKING** #4 (enhancement) Removed ember-template-compiler dependency and reduced payload size from 80kb to 19kb. This is a breaking change for people using custom templates as they now need to be compiled .hbs files instead of strings.  See "Using your own custom template" in README.md for a simple upgrade path.

# 1.0.0
*  First Official Release


addon/components/aupac-typeahead.js
import Ember from 'ember';
import footerTemplate from '../templates/components/aupac-typeahead/footer';
import headerTemplate from '../templates/components/aupac-typeahead/header';
import notFoundTemplate from '../templates/components/aupac-typeahead/not-found';
import pendingTemplate from '../templates/components/aupac-typeahead/pending';
import suggestionTemplate from '../templates/components/aupac-typeahead/suggestion';

const {observer, isNone, run, debug, getOwner, Component} = Ember;

const Key = {
  BACKSPACE : 8,
  DELETE : 46,
  ENTER : 13
};

export default Component.extend({
  init() {
    this._super(...arguments);
    let owner = getOwner(this);
    let lookup = owner.lookup('component:-typeahead-suggestion');
    if(isNone(lookup)){
      owner.register('component:-typeahead-suggestion', Component);
    }
  },

  //input tag attributes
  tagName : 'input',
  classNames: ['aupac-typeahead'],
  attributeBindings : ['disabled','placeholder', 'name', 'tabindex'],
  disabled : false, //@public
  placeholder : 'Search', //@public
  name : '', //@public

  //Actions
  action() {}, //@public
  selection : null, //@public
  source() {}, //@public

  //typeahead.js Customizations
  highlight: true, //@public
  hint: true, //@public
  minLength: 2, //@public
  typeaheadClassNames: {}, //@public
  autoFocus: false, //@public
  limit : 15, //@public
  async : false, //@public
  datasetName : '', //@public
  allowFreeInput: false, //@public

  //HtmlBars Templates
  suggestionTemplate,  //@public
  notFoundTemplate,  //@public
  pendingTemplate,  //@public
  headerTemplate,  //@public
  footerTemplate,  //@public

  //Private
  _typeahead: null,
  oldValue: null,


  /**
   * @public
   * @param selection - the item selected by the user
   * @returns {*}
   */
  display : function(selection) {
    return selection;
  },

  /**
   * @public
   * @param selection the item selected by the user
   */
  transformSelection(selection){
    return selection;
  },

  /**
   * @public
   * @param selection the item selected by the user
   */
  setValue : function(selection) {
    if (this.get('_typeahead')) { // Was failing in tests with this probably due to a stray observer
      selection = this.get('transformSelection')(selection);
      if(selection) {
        this.get('_typeahead').typeahead('val', selection);
      } else {
        this.get('_typeahead').typeahead('val', '');
      }
    }
  },

  didInsertElement: function () {
    this._super(...arguments);
    this.initializeTypeahead();
    if (this.get('autoFocus') === true) {
      this.get('_typeahead').focus();
    }
    this.addObserver('disabled', this.disabledStateChanged);
  },

  disabledStateChanged() {
    // Toggling the disabled attribute on the controller does not update the hint, need to do this manually.
    this.$().parent().find('input.tt-hint').prop('disabled', this.get('disabled'));
  },

  initializeTypeahead: function() {
    const self = this;
    //Setup the typeahead
    const t = this.$().typeahead({
      highlight: this.get('highlight'),
      hint: this.get('hint'),
      minLength: this.get('minLength'),
      classNames: this.get('typeaheadClassNames')
      }, {
        component : this,
        name: this.get('datasetName') || 'default',
        display: this.get('display'),
        async: this.get('async'),
        limit: this.get('limit'),
        source: this.get('source'),
        templates: {
          suggestion: function (model) {
            const el = document.createElement('div');
            if (typeof model !== 'object') {
              model = Ember.Object.create({
                displayName: model
              });
            }
            let owner = getOwner(self);
            let ComponentFactory = owner.factoryFor('component:-typeahead-suggestion');
            ComponentFactory.create({
              model,
              layout: self.get('suggestionTemplate'),
            }).appendTo(el);
            return el;
          },
          notFound: function (query) {
            const el = document.createElement('div');
            const model = Ember.Object.create({
              query
            });
            let owner = getOwner(self);
            let ComponentFactory = owner.factoryFor('component:-typeahead-suggestion');
            ComponentFactory.create({
              model,
              layout: self.get('notFoundTemplate'),
            }).appendTo(el);
            return el;
          },
          pending: function (query) {
            const el = document.createElement('div');
            const model = Ember.Object.create({
              query
            });
            let owner = getOwner(self);
            let ComponentFactory = owner.factoryFor('component:-typeahead-suggestion');
            ComponentFactory.create({
              model,
              layout: self.get('pendingTemplate')
            }).appendTo(el);
            return el;
          },
          header: function (query, suggestions) {
            const el = document.createElement('div');
            const model = Ember.Object.create({
              query,
              suggestions
            });
            let owner = getOwner(self);
            let ComponentFactory = owner.factoryFor('component:-typeahead-suggestion');
            ComponentFactory.create({
              model,
              layout: self.get('headerTemplate')
            }).appendTo(el);
            return el;
          },
          footer: function (query, suggestions) {
            const el = document.createElement('div');
            const model = Ember.Object.create({
              query,
              suggestions
            });
            let owner = getOwner(self);
            let ComponentFactory = owner.factoryFor('component:-typeahead-suggestion');
            ComponentFactory.create({
              model,
              layout: self.get('footerTemplate')
            }).appendTo(el);
            return el;
          }
        }
    });
    this.set('_typeahead', t);

    // Set selected object
    t.on('typeahead:autocompleted', run.bind(this, (jqEvent, suggestionObject /*, nameOfDatasetSuggestionBelongsTo*/) => {
        this.updateSelectionWhenChanged(suggestionObject, 'typeahead:autocompleted');
    }));

    t.on('typeahead:selected', run.bind(this, (jqEvent, suggestionObject /*, nameOfDatasetSuggestionBelongsTo*/) => {
      this.updateSelectionWhenChanged(suggestionObject, 'typeahead:selected');
    }));

    t.on('keyup', run.bind(this, (jqEvent) => {
      //Handle the case whereby the user presses the delete or backspace key, in either case
      //the selection is no longer valid.
      if (jqEvent.which === Key.BACKSPACE || jqEvent.which === Key.DELETE) {
        if (!this.get('allowFreeInput')) {
          debug("Removing model");
          const value = this.get('_typeahead').typeahead('val'); //cache value
          this.updateSelectionWhenChanged(null, 'keyup');
          this.setValue(value); // restore the text, thus allowing the user to make corrections
        }
      }
    }));

    t.on('keydown', run.bind(this, (jqEvent) => {
      if (jqEvent.which === Key.ENTER) {
        this._commitSelection('keydown');
      }
    }));

    t.on('focusout', run.bind(this, (/*jqEvent*/) => {
      //the user has now left the control, update display with current binding or reset to blank
      this._commitSelection('focusout');
    }));

  },

  _commitSelection: function(evt_name) {
    const model = this.get('selection');
    if (this.get('allowFreeInput')) {
      const value = this.get('_typeahead').typeahead('val');
      this.updateSelectionWhenChanged(value, evt_name);
    } else if (model) {
      this.setValue(model);
    } else {
      this.setValue(null);
    }
  },

  selectionUpdated: observer('selection', '_typeahead',function() {
    const selection = this.get('selection');
    if(isNone(selection)) {
      this.setValue(null);
    } else {
      this.setValue(selection);
    }
  }),

  updateSelectionWhenChanged: function(value, evt_name){
    const oldValue = this.get('oldValue');
    if (oldValue !== value) {
      this.set('oldValue', value);
      this.set('selection', value);
      this.sendAction('action', value, evt_name);
    }
  },

  willDestroyElement : function() {
    this._super(...arguments);
    let t = this.get('_typeahead');

    //Remove custom event handlers before destroying
    t.off('typeahead:autocompleted');
    t.off('typeahead:selected');
    t.off('keyup');
    t.off('keydown');
    t.off('focusout');

    t.typeahead('destroy');

    //Dereference the element
    this.set('_typeahead', null);
  }

});
